src/engraving/io/mscreader.cpp

```diff
@@ -0,0 +1,379 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-CLA-applies
+ *
+ * MuseScore
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2021 MuseScore BVBA and others
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include "mscreader.h"
+
+#include <QXmlStreamReader>
+#include <QFile>
+#include <QFileInfo>
+#include <QDir>
+#include <QDirIterator>
+
+#include "thirdparty/qzip/qzipreader_p.h"
+
+#include "log.h"
+
+//! NOTE The current implementation resolves files by extension.
+//! This will probably be changed in the future.
+
+using namespace mu::engraving;
+
+MscReader::MscReader(const Params& params)
+    : m_params(params)
+{
+}
+
+MscReader::~MscReader()
+{
+    close();
+}
+
+void MscReader::setParams(const Params& params)
+{
+    IF_ASSERT_FAILED(!isOpened()) {
+        return;
+    }
+
+    if (m_reader) {
+        delete m_reader;
+        m_reader = nullptr;
+    }
+
+    m_params = params;
+}
+
+const MscReader::Params& MscReader::params() const
+{
+    return m_params;
+}
+
+bool MscReader::open()
+{
+    return reader()->open(m_params.device, m_params.filePath);
+}
+
+void MscReader::close()
+{
+    if (m_reader) {
+        m_reader->close();
+
+        delete m_reader;
+        m_reader = nullptr;
+    }
+}
+
+bool MscReader::isOpened() const
+{
+    return m_reader->isOpened();

```

Comment:
might be safer to use "reader()" here

Reply:
I did just that to the writer ...
but I think differently
I'll fix it to:
```
return m_reader ? m_reader->isOpened() : false;
```